dag_in_context/src/greedy_dag_extractor.rs

```diff
@@ -537,13 +544,26 @@ pub fn extract_with_paths(
         }
 
         let sort_of_node = info.get_sort_of_eclass(classid);
-        if sort_of_node == "Expr"
-            && effectful_paths.is_some()
-            && effectful_paths.unwrap().contains_key(&rootid)
-        {
-            let effectful_nodes = effectful_paths.as_ref().unwrap().get(&rootid).unwrap();
-            if let Some(is_stateful) = extractor.is_node_effectful(nodeid.clone()) {
-                if is_stateful && !effectful_nodes.contains(&nodeid) {
+        // if node is effectful, we only consider it if it is in the effectful path
+        if sort_of_node == "Expr" && effectful_paths.is_some() {
```

Comment:
This is now getting really complex with 5 (!!) nested ifs. We need to work on this together and simplify the code. Also there are two `is_stateful` in the ifs and one is redundant.

Reply:
I agree

Reply:
refactored